cmd/ssh.go

```diff
@@ -244,3 +178,226 @@ func SSH(cmd *cobra.Command, args []string) error {
 		}
 	}
 }
+
+func handleSSHConnWithOpenVSCode(ctx context.Context, execCh chan types.Exec, isNew bool, sshArgsByKey map[string]string, errCh chan error) error {
+	for {
+		select {
+		case e := <-execCh:
+			if e.Status == types.StatusRunning {
+				ensureHosts(e)
+				defer cleanupHosts(e)
+				privKey := getPrivateKeyPathFromArgs(ctx, e, sshArgsByKey)
+
+				err := handleCopySourceDir(isNew, e, privKey)
+				if err != nil {
+					return err
+				}
+
+				ui.Infof("🔧 Setting up VS Code ...")
+				arg := fmt.Sprintf("vscode-remote://ssh-remote+%s@%s/home/ubuntu", e.Connection.User, e.Connection.Host)
+
+				codeCmd := exec.Command("code", "--folder-uri="+arg)
+				codeCmd.Stdout = os.Stdout
+				codeCmd.Stderr = os.Stderr
+				if e := codeCmd.Run(); e != nil {
+					ui.Errorf("Failed to start VS Code: %v", e)
+					os.Exit(1)
+				}
+				ui.Successf("✅ VS Code is ready!")
+				return nil
+			}
+
+		case err := <-errCh:
+			var e *types.Error
+			if errors.As(err, &e) {
+				uie := &ui.Error{Error: e}
+				fmt.Println(uie.Verbose())
+				os.Exit(1)
+			}
+			return err
+
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func cleanupHosts(e types.Exec) {
+	if err := ssh.RemoveHost("uw:" + e.ID); err != nil {
+		ui.Debugf("Failed to remove host from ssh config: %v", err)
+	}
+}
+
+func ensureHosts(e types.Exec) {
+	if e.Connection == nil {
+		ui.Errorf("❌ Something unexpected happened. No connection info found for session %q", e.ID)
+		ui.Infof("Run `unweave ls` to see the status of your session and try connecting manually.")
+		os.Exit(1)
+	}
+
+	ui.Infof("🚀 Session %q up and running", e.ID)
+
+	if err := ssh.RemoveKnownHostsEntry(e.Connection.Host); err != nil {
+		// Log and continue anyway. Most likely the entry is not there.
+		ui.Debugf("Failed to remove known_hosts entry: %v", err)
+	}
+
+	if err := ssh.AddHost("uw:"+e.ID, e.Connection.Host, e.Connection.User, e.Connection.Port); err != nil {
+		ui.Debugf("Failed to add host to ssh config: %v", err)
+	}
+}
+
+func handleCopySourceDir(isNew bool, e types.Exec, privKey string) error {
+	// TODO: Wait until port is open before cleaning up the source code
+
+	if !config.NoCopySource && isNew {
+		dir, err := config.GetActiveProjectPath()
+		if err != nil {
+			ui.Errorf("Failed to get active project path. Skipping copying source directory")
+			return fmt.Errorf("failed to get active project path: %v", err)
+		}
+		if err := copySource(e.ID, dir, "/home/ubuntu", *e.Connection, privKey); err != nil {
+			fmt.Println(err)
+		}
+	} else {
+		ui.Infof("Skipping copying source directory")
+	}
+	return nil
+}
```

Comment:
These private methods are shared codeblocks between the SSH flow and the Code flow once you have an Exec